### PR TITLE
Added new module avi_user.py

### PR DESCRIPTION
**Pull request by gitshrikant**
##### SUMMARY

- Added new module

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

- network/avi/avi_user.py

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Review by chaitanyaavi — APPROVED** [review]
+shipit

File: lib/ansible/modules/network/avi/avi_user.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    - This module can be used for creation, updation and deletion of a user.
+version_added: 2.9
+requirements: [ avisdk ]
+options:
+    state:
+        description:
+            - The state that should be applied on the entity.
+        default: present
+        choices: ["absent", "present"]
+    name:
```
Comment (Qalthos) [inline lib/ansible/modules/network/avi/avi_user.py:49]: Should have `reqiured: true`, along with `obj_username` and `obj_password` according to the module argspec

**Review by Qalthos — CHANGES_REQUESTED** [review]
This module is not likely to be accepted in its current format. Specifically, from https://docs.ansible.com/ansible/latest/dev_guide/developing_modules.html:

* Modules should not require that a user know all the underlying options of an API/tool to be used. For instance, if the legal values for a required module parameter cannot be documented, that’s a sign that the module would be rejected.
* Modules should typically encompass much of the logic for interacting with a resource. A lightweight wrapper around an API that does not contain much logic would likely cause users to offload too much logic into a playbook, and for this reason the module would be rejected. Instead try creating multiple modules for interacting with smaller individual pieces of the API.

A lot of these modules bend those rules pretty far, but directly using API paths as parameters is a pretty clear violation of the first, and the organization of the arguments as a whole makes me suspect the second. Ansible modules should not be a 1-to-1 mapping of an API, but a coherent set of tools to accomplish specific tasks that may happen to use an API.

**Comment by grastogi23** [comment]
> This module is not likely to be accepted in its current format. Specifically, from https://docs.ansible.com/ansible/latest/dev_guide/developing_modules.html:
> 
> * Modules should not require that a user know all the underlying options of an API/tool to be used. For instance, if the legal values for a required module parameter cannot be documented, that’s a sign that the module would be rejected.
> * Modules should typically encompass much of the logic for interacting with a resource. A lightweight wrapper around an API that does not contain much logic would likely cause users to offload too much logic into a playbook, and for this reason the module would be rejected. Instead try creating multiple modules for interacting with smaller individual pieces of the API.
> 
> A lot of these modules bend those rules pretty far, but directly using API paths as parameters is a pretty clear violation of the first, and the organization of the arguments as a whole makes me suspect the second. Ansible modules should not be a 1-to-1 mapping of an API, but a coherent set of tools to accomplish specific tasks that may happen to use an API.



> This module is not likely to be accepted in its current format. Specifically, from https://docs.ansible.com/ansible/latest/dev_guide/developing_modules.html:
> 
> * Modules should not require that a user know all the underlying options of an API/tool to be used. For instance, if the legal values for a required module parameter cannot be documented, that’s a sign that the module would be rejected.

As such this module does not violate that. The legal values are indeed documented here. In addition, ansible module exists in a context of a solution or a product. So, user cannot just use an ansible module without knowing the context of what it is. 

> * Modules should typically encompass much of the logic for interacting with a resource. A lightweight wrapper around an API that does not contain much logic would likely cause users to offload too much logic into a playbook, and for this reason the module would be rejected. Instead try creating multiple modules for interacting with smaller individual pieces of the API.

Whenever there has been a requirement of stitching together workflows the modules do that. In this case it is a first class REST resource and so it does warrant a proper ansible module. This allows customers to  have proper modularity. Having a single module do multiple things is not so nice in the same way as single API doing multiple REST resources.


> 
> A lot of these modules bend those rules pretty far, but directly using API paths as parameters is a pretty clear violation of the first, and the organization of the arguments as a whole makes me suspect the second. Ansible modules should not be a 1-to-1 mapping of an API, but a coherent set of tools to accomplish specific tasks that may happen to use an API.

Actually, I would say anytime an ansible module reduces the scope of the operations that is possible via API then it is sub-optimal. Also, not having consistent parameter names are confusing for the user. Not to mention documentation nightmare.  Eg. this is a module for Avi User and caller of this module needs to know what are these settings and how it maps to the settings on the product.

Many of our customers have actually thanked us and highlighted that they can do every thing with our modules that they would be able to do via API or UI. That is very powerful feature that should be encouraged for module development.

**Comment by Qalthos** [comment]
From today's IRC meeting:

> \<grastogi> is core of the debate about using syntax like "/api/tenant?name=admin"?
> \<Qalthos> grastogi: That is a lot of it
> \<grastogi> Yes, having that syntax has allowed us to properly check against the reference. In a way your argument is that name should be enough for matching the reference. However, having it this way allowed us to pass hints to the module about type of the reference. This we used to then create a generic implementation to match against what is returned from the Avi controller and not having to pass UUIDs in the playbooks.
> \<grastogi> yes, there is a path but that is mostly a reference and always in the format of /api/<obj_type>?name=<obj_name>
> \<Qalthos> Then surely the module could accept a name as a parameter and send the formatted path to the generic implementation
> \<Qalthos> or a dictionary of name and type if name alone is insufficient
> \<Qalthos> The user should not have to know or care how the paths are formatted, that is the point I am trying to make
> \<grastogi> yes, I completely agree with you. It was not our best option. We could not achieve the solution with just name. The other one is not as clean so didn't implement that. User does need to know the type of object as you can't put a name and not know whether it is a virtualservice or pool or tenant
> \<Qalthos> In that case, I would suggest adding suboptions to each of those parameters, which should allow you to be more explicit about what sorts of values are expected or appropriate

In short, a dictionary representing the name and type of each resource is more acceptable than an API path.

**Comment by grastogi23** [comment]
> From today's IRC meeting:
> 
> > <grastogi> is core of the debate about using syntax like "/api/tenant?name=admin"?
> > <Qalthos> grastogi: That is a lot of it
> > <grastogi> Yes, having that syntax has allowed us to properly check against the reference. In a way your argument is that name should be enough for matching the reference. However, having it this way allowed us to pass hints to the module about type of the reference. This we used to then create a generic implementation to match against what is returned from the Avi controller and not having to pass UUIDs in the playbooks.
> > <grastogi> yes, there is a path but that is mostly a reference and always in the format of /api/<obj_type>?name=<obj_name>
> > <Qalthos> Then surely the module could accept a name as a parameter and send the formatted path to the generic implementation
> > <Qalthos> or a dictionary of name and type if name alone is insufficient
> > <Qalthos> The user should not have to know or care how the paths are formatted, that is the point I am trying to make
> > <grastogi> yes, I completely agree with you. It was not our best option. We could not achieve the solution with just name. The other one is not as clean so didn't implement that. User does need to know the type of object as you can't put a name and not know whether it is a virtualservice or pool or tenant
> > <Qalthos> In that case, I would suggest adding suboptions to each of those parameters, which should allow you to be more explicit about what sorts of values are expected or appropriate
> 
> In short, a dictionary representing the name and type of each resource is more acceptable than an API path.

Hi @Qalthos,

I wasn't able to get the issue across properly. The Avi APIs need a reference to the object. The reference is a URL REST reference to the object eg.  http://avi_endpoint/tenant/tenant-uuid. 

The references are similar to the Amazon's ARN or google firewall's network parameter. Eg. ssl_certificate_id: "arn:aws:iam::123456789012:server-certificate/company/servercerts/ProdServerCert" from the Aws module. 

However, we made it simpler for the user to give partial URL path rather than the whole reference URL. If you were to compare against aws or google we could have instead used  tenant/admin instead of "/api/tenant?name=admin". We went with keeping semantics of API and Ansible same so users don't get confused.

As far as on AVI module side we cannot send a tenant name simply when expectation of API is to receive a full reference without having to re-write all of API code. 

We would like you to reconsider in light of what this parameter actually means. It is not a name but a reference. Also, existing modules already work with if someone passes full reference url rather than the filter semantics. 

```
- local_action:
    module: ec2_elb_lb
    name: "test-please-delete"
    state: present
    zones:
      - us-east-1a
      - us-east-1d
    listeners:
      - protocol: http # options are http, https, ssl, tcp
        load_balancer_port: 80
        instance_port: 80
        proxy_protocol: True
      - protocol: https
        load_balancer_port: 443
        instance_protocol: http # optional, defaults to value of protocol setting
        instance_port: 80
        # ssl certificate required for https or ssl
        ssl_certificate_id: "arn:aws:iam::123456789012:server-certificate/company/servercerts/ProdServerCert"
```
Here are more examples 
https://github.com/ansible/ansible/blob/4bf051b8bb3877f95ce1c807000aec1b2929aee0/lib/ansible/modules/cloud/google/gcp_compute_firewall.py#L138

https://github.com/ansible/ansible/blob/f6a0f9874ddeb4b79012272eb5a4a92c0da76756/lib/ansible/modules/cloud/amazon/ec2_elb_lb.py#L179

thanks,
Gaurav

**Comment by grastogi23** [comment]
@gundalow requesting your comments as well.

**Review by grastogi23 — APPROVED** [review]
+shipit

**Review by Qalthos — APPROVED** [review]

**Pull request merged**